### PR TITLE
Update to multer 1.3.1 and automatically clean uploaded files

### PR DESCRIPTION
**Pull request by JedWatson**
This has been frustrating me for a long time, but I've worked out a way to update `multer` to the latest version and retain backwards compatibility while still being safe with uploaded files.

This resolves several issue that have been raised about our use of an old multer version, including #4694 #4428 #3597 #3175 and #1563

It also introduces some new features:

- Keystone will automatically delete uploaded files at the end of the request
- You can disable the built-in multer in your app by setting the `handle uploads` Keystone option to `false` (it will still be included by the admin UI)
- You can pass options to multer my setting the `multer options` Keystone option

e.g

```
keystone.init({
  // ...
  'handle uploads': false, // turns it off
});

// or

keystone.init({
  // ...
  'multer options': { /* passed to multer */ },
});
```

**Technically** this is possibly a breaking change, because we now automatically clean up uploaded files. This was never a thing I'd expect to be used but any weird edge-case implementation that expects uploaded files to hang around after the request is completed will break. Because of this, I'd like to merge this for `4.0.0` final (not RC) and make a prominent note about it in the changelog.

### CODE REVIEW

**Comment by intrnet** [comment]
I've tried on 4.0.0-rc.1 and 4.0.0-beta.8 version but I met same error, every requests ends without respond. And I guess there is little mistake in **_server/bindBodyParser.js_** `module.exports = function bundBodyParser (keystone, app) {` I think there should be `module.exports = function bindBodyParser (keystone, app) {` but this is not the couse of the problem.

**Comment by JedWatson** [comment]
Thanks for catching that @intrnet 

Just to confirm, do you mean you tested merging this into your own branch based on `4.0.0-rc.1`?

We're looking into the failure to connect now. It all worked for me in local testing, but it looks like there is an issue with this in different environments.

**Comment by intrnet** [comment]
I simply update multer and selected files, according to your changes ([Update to multer 1.3.1 and automatically clean uploaded files](https://github.com/keystonejs/keystone/pull/4704/commits/a3fcb46e791f4ed75c6cb06c82535ce4eecdce61)) in local testing based on linux ubuntu.

**Comment by stennie** [comment]
@JedWatson I'm encountering the same issue with e2e tests (and keystone-demo) as reported in CI failures: `/keystone` times out waiting for response. Tried with Node 6.14.3 and 8.11.3. Also tried with both `yarn` and `npm` since your PR includes `yarn.lock` (I believe Travis is using `npm`).

I'm pulling from the `update-multer` branch associated with this PR, so not sure what is different in the test environments. Is there a specific `yarn`/`npm` incantation you would suggest to ensure a clean rebuild?

File: lib/uploads.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+var fs = require('fs');
+var multer = require('multer');
+var os = require('os');
+
+function handleUploadedFiles (req, res, next) {
+	if (!req.files || !Array.isArray(req.files)) return;
```
Comment (ttsirkia) [inline lib/uploads.js]: Should this be `return next();`?

**Comment by ttsirkia** [comment]
I like the approach of cleaning the files when the request ends. However, if someone relies on the files still being on the disk, would it be useful to add a possibilty for a flag in the request or response object which skips the deletion for that request? Or should we assume here that if the file is wanted to store permanently, it must be copied elsewhere?

**Pull request merged**